### PR TITLE
[SPARK-5182] [SPARK-5528] [SPARK-5509] [SPARK-3575] [SQL] Parquet data source improvements

### PR DESCRIPTION
**Pull request by liancheng**
This PR adds three major improvements to Parquet data source:
1. Partition discovery
   
   While reading Parquet files resides in Hive style partition directories, `ParquetRelation2` automatically discovers partitioning information and infers partition column types.
   
   This is also a partial work for [SPARK-5182](https://issues.apache.org/jira/browse/SPARK-5182), which aims to provide first class partitioning support for the data source API.  Related code in this PR can be easily extracted to the data source API level in future versions.
2. Schema merging
   
   When enabled, Parquet data source collects schema information from all Parquet part-files and tries to merge them.  Exceptions are thrown when incompatible schemas are detected.  This feature is controlled by data source option `parquet.mergeSchema`, and is enabled by default.

3. Metastore Parquet table conversion moved to analysis phase
   
   This greatly simplifies the conversion logic.  `ParquetConversion` strategy can be removed once the old Parquet implementation is removed in the future.

This version of Parquet data source aims to entirely replace the old Parquet implementation.  However, the old version hasn't been removed yet.  Users can fall back to the old version by turning off SQL configuration `spark.sql.parquet.useDataSourceApi`.

Other JIRA tickets fixed as side effects in this PR:
- [SPARK-5509](https://issues.apache.org/jira/browse/SPARK-5509): `EqualTo` now uses a proper `Ordering` to compare binary types.
- [SPARK-3575](https://issues.apache.org/jira/browse/SPARK-3575): Metastore schema is now preserved and passed to `ParquetRelation2` via data source option `parquet.metastoreSchema`.

TODO:
- [ ] More test cases for partition discovery
- [x] Fix write path after data source write support (#4294) is merged
  
      It turned out to be non-trivial to fall back to old Parquet implementation on the write path when Parquet data source is enabled.  Since we're planning to include data source write support in 1.3.0, I simply ignored two test cases involving Parquet insertion for now.

- [ ] Fix outdated comments and documentations

PS: This PR looks big, but more than a half of the changed lines in this PR are trivial changes to test cases. To test Parquet with and without the new data source, almost all Parquet test cases are moved into wrapper driver functions. This introduces hundreds of lines of changes.

[Review on Reviewable](https://reviewable.io/reviews/apache/spark/4308)

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 1]
```diff
@@ -175,7 +175,10 @@ case class EqualTo(left: Expression, right: Expression) extends BinaryComparison
       null
     } else {
       val r = right.eval(input)
-      if (r == null) null else l == r
+      if (r == null) null
+      else if (left.dataType != BinaryType) l == r
+      else BinaryType.ordering.compare(
+        l.asInstanceOf[Array[Byte]], r.asInstanceOf[Array[Byte]]) == 0
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: This fixes SPARK-5509. Hit this bug while testing Parquet filters for new data source implementation.

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTest.scala [hunk 2]
```diff
@@ -89,7 +89,7 @@ trait ParquetTest {
       (data: Seq[T])
       (f: String => Unit): Unit = {
     withTempPath { file =>
-      sparkContext.parallelize(data).saveAsParquetFile(file.getCanonicalPath)
+      sparkContext.parallelize(data, 3).saveAsParquetFile(file.getCanonicalPath)
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTest.scala]: Oops, debugging code, should revert this change.

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala [hunk 3]
```diff
@@ -81,116 +101,158 @@ private[parquet] case class Partition(partitionValues: Map[String, Any], files:
...
 
-    val selectedPartitions =
-      if (partitionKeys.nonEmpty && predicates.nonEmpty) {
-        partitions.filter { part =>
-          inputData(0) = part.partitionValues.values.head
-          pruningCondition(inputData)
-        }
-      } else {
-        partitions
+  // This is mostly a hack so that we can use the existing parquet filter code.
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala]: Should stop using Catalyst expressions and remove `CatalystScan` after lifting partition discovery and partition pruning to data source API level.

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala [hunk 4]
```diff
@@ -227,66 +294,302 @@ case class ParquetRelation2(path: String)(@transient val sqlContext: SQLContext)
...
+  /**
+   * Parses a single partition, returns column names and values of each partition column.  For
+   * example, given:
+   * {{{
+   *   basePath = hdfs://<host>:<port>/path/to/partition
+   *   partitionPath = hdfs://<host>:<port>/path/to/partition/a=42/b=hello/c=3.14
+   * }}}
+   * it returns:
+   * {{{
+   *   PartitionDesc(
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala]: Typo, should be `PartitionValues`.

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala [hunk 5]
```diff
@@ -227,66 +294,302 @@ case class ParquetRelation2(path: String)(@transient val sqlContext: SQLContext)
...
+
+    PartitionSpec(StructType(fields), partitions)
+  }
+
+  /**
+   * Parses a single partition, returns column names and values of each partition column.  For
+   * example, given:
+   * {{{
+   *   basePath = hdfs://<host>:<port>/path/to/partition
+   *   partitionPath = hdfs://<host>:<port>/path/to/partition/a=42/b=hello/c=3.14
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala]: Fix outdated comments.

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala [hunk 6]
```diff
@@ -227,66 +294,302 @@ case class ParquetRelation2(path: String)(@transient val sqlContext: SQLContext)
...
+    // Column names of all partitions must match
+    assert(distinctColNamesOfPartitions.size == 1, {
+      val list = distinctColNamesOfPartitions.mkString("\t", "\n", "")
+      s"Conflicting partition column names detected:\n$list"
+    })
+
+    // Resolves possible type conflicts for each column
+    val resolvedValues = (0 until columnCount).map { i =>
+      resolveTypeConflicts(descs.map(_.literals(i)))
+    }
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala]: Should check and throw if some column data type ends up with `NullType`.

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala [hunk 7]
```diff
@@ -227,66 +294,302 @@ case class ParquetRelation2(path: String)(@transient val sqlContext: SQLContext)
...
+    // First tries integral types
+    Try(Literal(Integer.parseInt(raw), IntegerType))
+      .orElse(Try(Literal(JLong.parseLong(raw), LongType)))
+      // Then falls back to fractional types
+      .orElse(Try(Literal(JFloat.parseFloat(raw), FloatType)))
+      .orElse(Try(Literal(JDouble.parseDouble(raw), DoubleType)))
+      .orElse(Try(Literal(new JBigDecimal(raw), DecimalType.Unlimited)))
+      // Then falls back to string
+      .getOrElse {
+        if (raw == defaultPartitionName) Literal(null, NullType) else Literal(raw, StringType)
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala]: For string, should we convert default partition name to `null` or empty string? Need to check Hive behaviour here.

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTypes.scala [hunk 8]
```diff
@@ -490,4 +489,55 @@ private[parquet] object ParquetTypesConverter extends Logging {
       attributes
     }
   }
+
+  def mergeCatalystSchemas(left: StructType, right: StructType): StructType =
+    mergeCatalystDataTypes(left, right).asInstanceOf[StructType]
+
+  def mergeCatalystDataTypes(left: DataType, right: DataType): DataType =
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTypes.scala]: would be great to add more comment explaining what's going on

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala (hunk 1 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: btw this is really expensive. i'd use sth like this: http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/primitives/UnsignedBytes.html

If you don't want to change it as part of this PR, file a jira ticket to track it.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 9]
```diff
@@ -262,8 +262,14 @@ class SQLContext(@transient val sparkContext: SparkContext)
    *
    * @group userf
    */
-  def parquetFile(path: String): DataFrame =
-    DataFrame(this, parquet.ParquetRelation(path, Some(sparkContext.hadoopConfiguration), this))
+  @scala.annotation.varargs
+  def parquetFile(paths: String*): DataFrame =
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: as commented on the other pr, use
```scala
def parquetFile(path: String, paths: String*): DataFrame
```

to make sure this is not ambiguous if we overload the function with another varargs

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: Thanks. Makes sense.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala (hunk 1 above)
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: Filed SPARK-5553 to track this. I'd like to make sure equality comparison for binary types works properly in this PR. Also, we're already using `Ordering` to compare binary values in `LessThan` and `GreaterThan` etc., so at least this isn't a regression.

**Comment by liancheng** [comment]
retest this please

**Comment by liancheng** [comment]
The last build failure was caused by a flaky ML test case, which is now fixed in master.

**Comment by liancheng** [comment]
retest this please.

The last build failure reports that `isFile` and `isRoot` are not member of `org.apache.hadoop.fs.FileStatus`, which doesn't make sense (the pull request builder uses Hadoop 2.3.0, and these methods are definitely defined in `FileStatus`).

**Comment by liancheng** [comment]
retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 10]
```diff
@@ -263,8 +263,14 @@ class SQLContext(@transient val sparkContext: SparkContext)
    *
    * @group userf
    */
-  def parquetFile(path: String): DataFrame =
-    DataFrame(this, parquet.ParquetRelation(path, Some(sparkContext.hadoopConfiguration), this))
+  @scala.annotation.varargs
+  def parquetFile(path: String, paths: String*): DataFrame =
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: why not just `path: String*`?

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTypes.scala (hunk 8 above)
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTypes.scala]: also should this live in catalyst?  Seems generally useful.

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala [hunk 11]
```diff
@@ -81,117 +123,189 @@ private[parquet] case class Partition(partitionValues: Map[String, Any], files:
...
+    (paths: Seq[String], parameters: Map[String, String], maybeSchema: Option[StructType] = None)
+    (@transient val sqlContext: SQLContext)
+  extends CatalystScan
+  with InsertableRelation
+  with SparkHadoopMapReduceUtil
+  with Logging {
+
+  // Should we merge schemas from all Parquet part-files?
+  private val shouldMergeSchemas =
+    parameters.getOrElse(ParquetRelation2.MERGE_SCHEMA, "true").toBoolean
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala]: I think this should be a datasource option instead of a global configuration.

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala]: oh sorry, it is.

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTypes.scala (hunk 8 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTypes.scala]: Yeah, will move it to Catalyst in follow-up PRs.

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala [hunk 12]
```diff
@@ -228,66 +347,397 @@ case class ParquetRelation2(path: String)(@transient val sqlContext: SQLContext)
...
+    sqlContext.sparkContext.runJob(data.queryExecution.executedPlan.execute(), writeShard _)
+    jobCommitter.commitJob(jobTaskContext)
+
+    metadataCache.refresh()
+  }
+}
+
+object ParquetRelation2 {
+  // Whether we should merge schemas collected from all Parquet part-files.
+  val MERGE_SCHEMA = "parquet.mergeSchema"
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala]: why prefix these with `parquet`? that seems redudant since you can only use them after specifying `USING org.apache.spark.sql.parquet`

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala (hunk 10 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: This is per @rxin's [this comment] [1], which I think makes sense.

[1]: https://github.com/apache/spark/pull/4308#discussion_r23983783

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala (hunk 12 above)
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala]: should we also have an option to turn off caching?

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala (hunk 9 above)
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: Are we actually ever going to do that for this function?  This makes it harder to do something like `parquetFile(listOfFiles: _*)` which I think is actually a common usecase.

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: i'll add that if we ever do overload this we can do this disambiguation then.

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: okay and i convinced @rxin too :)

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala (hunk 12 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala]: Thanks. Will address these in follow-up PR(s).

**Comment by liancheng** [comment]
Rebased (for the 8th time during the last 72 hours), should be ready to go once Jenkins nods. Will address comments in follow-up PRs.

**Comment by liancheng** [comment]
OK, rebased for the 9th time... Addressed all comments except for adding option to disable metadata caching, which I'd like to include in another PR.

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala [hunk 13]
```diff
@@ -228,66 +355,400 @@ case class ParquetRelation2(path: String)(@transient val sqlContext: SQLContext)
...
+    // Fills resolved literals back to each partition
+    values.zipWithIndex.map { case (d, index) =>
+      d.copy(literals = resolvedValues.map(_(index)))
+    }
+  }
+
+  /**
+   * Converts a string to a `Literal` with automatic type inference.  Currently only supports
+   * [[IntegerType]], [[LongType]], [[FloatType]], [[DoubleType]], [[DecimalType.Unlimited]], and
+   * [[StringType]].
```
Comment (nugend) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala]: Would it be reasonable to support DateType and *then* StringType. In my experience, breaking data down by a date partition is pretty common and useful. My thinking is that if you see a string in the format YYYY-MM-DD (for example, I do recommend that format given its alphabetical sorting, personally, but it doesn't have to be that), then you can *probably* safely assume that the partition is intended to be a date.

I'm not super familiar with this code though, so I'll have to defer to others' expertise.

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala]: Good point. Trying `DataType` before `StringType` makes sense. I can add this. Thanks for the suggestion!
